mangaki/mangaki/tests/test_anidb.py

```diff
@@ -47,5 +52,19 @@ def test_anidb_get(self):
             content_type='application/xml'
         )
         anime = self.anidb.get_or_update_work(11606)
-        self.assertNotEqual(anime.title, '')
-        self.assertNotEqual(anime.worktitle_set.count(), 0)
+
+        self.assertEqual(anime.title, 'Sangatsu no Lion')
+        self.assertEqual(anime.nb_episodes, 22)
+        self.assertEqual(anime.studio.title, 'Shaft')
+
+        author_ids = [q.artist_id for q in Staff.objects.filter(work_id=anime.pk, role_id=3)]
+        author_names = [Artist.objects.get(pk=at_id).name for at_id in author_ids]
+        self.assertEqual(author_names, ['Umino Chika'])
```

Comment:
I looked up your recommandations to improve queries to the database in `test_anidb.py`. I tried a few solutions scrolling down your comments, and thought it would be interesting to write a bit about it ! But yeah, the last recommandation is indeed the best, as expected since it reduces redundancy of queries and makes use of relationships between models.

*Note : woops, this text might be longer than I had anticipated when starting to write. Sorry !*


#### My first solution basically made too many queries if we are facing a big staff :
- 1 query to get the list of IDs of staff corresponding to the work id and the role id (so 3 queries in total)
- 1 query per ID of staff, in order to retrieve the name of that staff

Therefore, we basically add up to `(3 + 3*N)*A` queries where `N` is the size of the staff and `A` the number of animes.

```py
author_ids = [q.artist_id for q in Staff.objects.filter(work_id=anime.pk, role_id=3)]
author_names = [Artist.objects.get(pk=at_id).name for at_id in author_ids]

composer_ids = [q.artist_id for q in Staff.objects.filter(work_id=anime.pk, role_id=4)]
composer_names = [Artist.objects.get(pk=cp_id).name for cp_id in composer_ids]

director_ids = [q.artist_id for q in Staff.objects.filter(work_id=anime.pk, role_id=5)]
director_names = [Artist.objects.get(pk=dt_id).name for dt_id in director_ids]
```

#### I tried out a second solution very similar to the first one bug using Django's lookups (in this case, `pk__in`) as you mentioned in a previous comment. This makes it more easily readable by using role slugs and slightly reduces the count of queries if we are facing a big staff :
- Still 1 query to get the list of IDs of staff corresponding to the work id and the role id (so 3 queries total)
- But only 1 query to retrieve the names of the staff given their ID (so 3 queries total)

This basically gets us to `6A` queries where `A` the number of animes.

```py
author_ids = Staff.objects.filter(work_id=anime.pk, role__slug='author').values_list('artist_id', flat=True)
author_names = Artist.objects.filter(pk__in=author_ids).values_list('name', flat=True)

composer_ids = Staff.objects.filter(work_id=anime.pk, role__slug='composer').values_list('artist_id', flat=True)
composer_names = Artist.objects.filter(pk__in=composer_ids).values_list('name', flat=True)

director_ids = Staff.objects.filter(work_id=anime.pk, role__slug='director').values_list('artist_id', flat=True)
director_names = Artist.objects.filter(pk__in=director_ids).values_list('name', flat=True)
```

#### Now, digging up Django's documentation a bit more, I found out that I could bring those 2 queries to a single one per role using relationships between models. It basically means 3 queries per work id, so `3A` queries in total. Moreover, queries are way faster since I now make real use of relationships, which I didn't do until now :

```py
author_names = Staff.objects.filter(work_id=anime.pk, role__slug='author').values_list('artist__name', flat=True)
composer_names = Staff.objects.filter(work_id=anime.pk, role__slug='composer').values_list('artist__name', flat=True)
director_names = Staff.objects.filter(work_id=anime.pk, role__slug='director').values_list('artist__name', flat=True)
```

#### However, I'm still doing the same query for each role. In order to further reduce computation time, it is indeed better to retrieve the list of staff beforehand and then only pick out the artist names corresponding to a specific role thanks to a `filter`. This brings us down to 1 query per work id and just filtering afterwards :

```py
staff = Work.objects.get(pk=anime.pk).staff_set.all() # Only called once per work !

author_names = staff.filter(role__slug='author').values_list('artist__name', flat=True)
composer_names = staff.filter(role__slug='composer').values_list('artist__name', flat=True)
director_names = staff.filter(role__slug='director').values_list('artist__name', flat=True)
```

I don't see a possibility to further reduce computation time and the number of queries at the moment (given there is one ...). I tried to loop these in order to see how it affects computation time at a bigger scale. Here were the results for 5000 iterations :

- **First solution :** 22.339s (first run) ; 23.364s (second run) ; 22.824s (third run)
- **Second solution :** 21.259s (first run) ; 21.381s (second run) ; 21.452s (third run)
- **Third solution :** 15.063s (first run) ; 14.917s (second run) ; 15.338s (third run)
- **Fourth solution :** 13.246s (first run) ; 13.277s (second run) ; 13.252s (third run)